diff --git a/Makefile b/Makefile
index 5a51d51..ed2a817 100644
--- a/Makefile
+++ b/Makefile
@@ -19,17 +19,17 @@
 SHELL = /bin/sh
 
 # The C compiler and linker to use
-CC	= gcc
+CC	= x86_64-os_2-gcc
 CFLAGS	= -g -O2 -Wall
-LD	= gcc
+LD	= $(CC)
 LDFLAGS =
 
 # Feature flags:
 #   define TLF_FONTS to use TOIlet TLF fonts
-XCFLAGS	= -DTLF_FONTS
+XCFLAGS	=
 
 # Where to install files
-prefix	= /usr/local
+prefix	= /usr
 
 # Where the executables should be put
 BINDIR	= $(prefix)/bin
