From 6b086801b1b0cd3837a473d1feabf997e2e7472f Mon Sep 17 00:00:00 2001
From: ColeTrammer <cole.trammer@gmail.com>
Date: Sun, 19 Feb 2023 01:25:43 -0800
Subject: [PATCH 2/3] Pretend Iros supports shared libraries in libtool.m4

---
 libtool.m4 | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/libtool.m4 b/libtool.m4
index b92e284d9f9..3bd049edfc0 100644
--- a/libtool.m4
+++ b/libtool.m4
@@ -2435,6 +2435,17 @@ irix5* | irix6* | nonstopux*)
   hardcode_into_libs=yes
   ;;
 
+iros*)
+  version_type=linux
+  need_lib_prefix=no
+  need_version=no
+  library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${release}${shared_ext}$major $libname${shared_ext}'
+  soname_spec='${libname}${release}${shared_ext}$major'
+  shlibpath_var=LD_LIBRARY_PATH
+  shlibpath_overrides_runpath=yes
+  hardcode_into_libs=yes
+  ;;
+
 # No shared lib support for Linux oldld, aout, or coff.
 linux*oldld* | linux*aout* | linux*coff*)
   dynamic_linker=no
@@ -3104,6 +3115,10 @@ irix5* | irix6* | nonstopux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+iros*)
+  lt_cv_deplibs_check_method=pass_all
+  ;;
+
 # This must be Linux ELF.
 linux* | k*bsd*-gnu | kopensolaris*-gnu | uclinuxfdpiceabi)
   lt_cv_deplibs_check_method=pass_all
-- 
2.34.1

