diff --git a/Makefile b/Makefile
index 5a51d51..71ff19b 100644
--- a/Makefile
+++ b/Makefile
@@ -19,26 +19,26 @@
 SHELL = /bin/sh
 
 # The C compiler and linker to use
-CC	= gcc
+CC	= x86_64-os_2-gcc
 CFLAGS	= -g -O2 -Wall
-LD	= gcc
+LD	= x86_64-os_2-gcc
 LDFLAGS =
 
 # Feature flags:
 #   define TLF_FONTS to use TOIlet TLF fonts
-XCFLAGS	= -DTLF_FONTS
+XCFLAGS	=
 
 # Where to install files
-prefix	= /usr/local
+prefix	= $(ROOT)/ports/figlet
 
 # Where the executables should be put
-BINDIR	= $(prefix)/bin
+BINDIR	= $(prefix)/figlet/bin
 
 # Where the man page should be put
 MANDIR	= $(prefix)/man
 
 # Where figlet will search first for fonts (the ".flf" files).
-DEFAULTFONTDIR = $(prefix)/share/figlet
+DEFAULTFONTDIR = /share/figlet
 # Use this definition if you can't put things in $(prefix)/share/figlet
 #DEFAULTFONTDIR = fonts
 
@@ -54,12 +54,12 @@ DEFAULTFONTFILE = standard
 
 VERSION	= 2.2.5
 DIST	= figlet-$(VERSION)
-OBJS	= figlet.o zipio.o crc.o inflate.o utf8.o
+OBJS	= figlet.o zipio.o crc.o inflate.o
 BINS	= figlet chkfont figlist showfigfonts
 MANUAL	= figlet.6 chkfont.6 figlist.6 showfigfonts.6
-DFILES	= Makefile Makefile.tc $(MANUAL) $(OBJS:.o=.c) chkfont.c getopt.c \
+DFILES	= Makefile Makefile.tc $(MANUAL) $(OBJS:.o=.c) chkfont.c \
 	  figlist showfigfonts CHANGES FAQ README LICENSE figfont.txt \
-	  crc.h inflate.h zipio.h utf8.h run-tests.sh figmagic
+	  crc.h inflate.h zipio.h run-tests.sh figmagic
 
 .c.o:
 	$(CC) -c $(CFLAGS) $(XCFLAGS) -DDEFAULTFONTDIR=\"$(DEFAULTFONTDIR)\" \
@@ -78,13 +78,10 @@ clean:
 
 install: all
 	mkdir -p $(DESTDIR)$(BINDIR)
-	mkdir -p $(DESTDIR)$(MANDIR)/man6
-	mkdir -p $(DESTDIR)$(DEFAULTFONTDIR)
+	mkdir -p $(DESTDIR)$(prefix)/figlet$(DEFAULTFONTDIR)
 	cp $(BINS) $(DESTDIR)$(BINDIR)
-	cp $(MANUAL) $(DESTDIR)$(MANDIR)/man6
-	cp fonts/*.flf $(DESTDIR)$(DEFAULTFONTDIR)
-	cp fonts/*.flc $(DESTDIR)$(DEFAULTFONTDIR)
-
+	cp fonts/*.flf $(DESTDIR)$(prefix)/figlet$(DEFAULTFONTDIR)
+	cp fonts/*.flc $(DESTDIR)$(prefix)/figlet$(DEFAULTFONTDIR)
 dist:
 	rm -Rf $(DIST) $(DIST).tar.gz
 	mkdir $(DIST)/
