commit b4da92f75aec7acf7772736ef666369b677192d1
Author: ColeTrammer <cole.trammer@gmail.com>
Date:   Sun Jul 3 16:56:51 2022 +0000

    fake run automake

diff --git a/ld/Makefile.in b/ld/Makefile.in
index 61e93eea..cc6fbbb1 100644
--- a/ld/Makefile.in
+++ b/ld/Makefile.in
@@ -774,6 +774,7 @@ ALL_EMULATION_SOURCES = \
 	eelf_i386_haiku.c \
 	eelf_i386_ldso.c \
 	eelf_i386_sol2.c \
+	eelf_i386_iros.c \
 	eelf_i386_vxworks.c \
 	eelf_iamcu.c \
 	eelf_s390.c \
@@ -954,6 +955,7 @@ ALL_64_EMULATION_SOURCES = \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_haiku.c \
 	eelf_x86_64_sol2.c \
+	eelf_x86_64_iros.c \
 	ehppa64linux.c \
 	ei386pep.c \
 	emmo.c
