diff --git a/gcc/config/os_2.h b/gcc/config/os_2.h
index 509f29e8d..f3f1b6133 100644
--- a/gcc/config/os_2.h
+++ b/gcc/config/os_2.h
@@ -2,16 +2,20 @@
 #define TARGET_OS_2 1
 
 #undef LIB_SPEC
-#define LIB_SPEC "-lc"
+#define LIB_SPEC "%{pthread:-lpthread} %{!shared:-lc}"
 
 #undef STARTFILE_SPEC
-#define STARTFILE_SPEC "crt0.o%s crti.o%s crtbegin.o%s"
+#define STARTFILE_SPEC "%{!shared:crt0.o%s} crti.o%s %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
 
 #undef ENDFILE_SPEC
-#define ENDFILE_SPEC "crtend.o%s crtn.o%s"
+#define ENDFILE_SPEC "%{!shared:crtend.o%s} %{shared:crtendS.o%s} crtn.o%s"
+
+#define DYNAMIC_LINKER "/lib/loader.so"
 
 #undef LINK_SPEC
-#define LINK_SPEC "-z max-page-size=4096"
+#define LINK_SPEC                                                                                                         \
+    "%{shared:-shared} %{static:-static} -z max-page-size=4096 %{!shared:%{!static:-dynamic-linker " DYNAMIC_LINKER "}} " \
+    "%{!shared: %{!static: %{rdynamic:-export-dynamic}}}"
 
 #undef TARGET_OS_CPP_BUILTINS
 #define TARGET_OS_CPP_BUILTINS()        \
diff --git a/libstdc++-v3/config/os/generic/os_defines.h b/libstdc++-v3/config/os/generic/os_defines.h
index 4e08d448a..68991a27f 100644
--- a/libstdc++-v3/config/os/generic/os_defines.h
+++ b/libstdc++-v3/config/os/generic/os_defines.h
@@ -36,6 +36,6 @@
 // Disable the weak reference logic in gthr.h for os/generic because it
 // is broken on every platform unless there is implementation specific
 // workaround in gthr-posix.h and at link-time for static linking.
-#define _GLIBCXX_GTHREAD_USE_WEAK 0
+#define _GLIBCXX_GTHREAD_USE_WEAK 1
 
 #endif
