diff --git a/fixincludes/mkfixinc.sh b/fixincludes/mkfixinc.sh
index a52333b2a..490dc4db3 100755
--- a/fixincludes/mkfixinc.sh
+++ b/fixincludes/mkfixinc.sh
@@ -14,7 +14,7 @@ case $machine in
     i?86-*-cygwin* | \
     i?86-*-mingw32* | \
     x86_64-*-mingw32* | \
-    *os_2-* | \
+    *-os_2* | \
     powerpc-*-eabisim* | \
     powerpc-*-eabi*    | \
     powerpc-*-rtems*   | \
diff --git a/gcc/config.gcc b/gcc/config.gcc
index d7afa9b0b..893c7a574 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1497,6 +1497,8 @@ i[34567]86-*-os_2*)
     ;;
 x86_64-*-os_2*)
     tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h glibc-stdint.h i386/i386elf.h i386/x86-64.h os_2.h"
+    tmake_file="${tmake_file} t-slibgcc"
+	extra_options="${extra_options} gnu-user.opt"
     ;;
 hppa[12]*-*-hpux10*)
 	case ${target} in
diff --git a/gcc/config/os_2.h b/gcc/config/os_2.h
index 509f29e8d..f3f1b6133 100644
--- a/gcc/config/os_2.h
+++ b/gcc/config/os_2.h
@@ -2,16 +2,20 @@
 #define TARGET_OS_2 1
 
 #undef LIB_SPEC
-#define LIB_SPEC "-lc"
+#define LIB_SPEC "%{pthread:-lpthread} %{!shared:-lc}"
 
 #undef STARTFILE_SPEC
-#define STARTFILE_SPEC "crt0.o%s crti.o%s crtbegin.o%s"
+#define STARTFILE_SPEC "%{!shared:crt0.o%s} crti.o%s %{!shared:crtbegin.o%s} %{shared:crtbeginS.o%s}"
 
 #undef ENDFILE_SPEC
-#define ENDFILE_SPEC "crtend.o%s crtn.o%s"
+#define ENDFILE_SPEC "%{!shared:crtend.o%s} %{shared:crtendS.o%s} crtn.o%s"
+
+#define DYNAMIC_LINKER "/lib/loader.so"
 
 #undef LINK_SPEC
-#define LINK_SPEC "-z max-page-size=4096"
+#define LINK_SPEC                                                                                                         \
+    "%{shared:-shared} %{static:-static} -z max-page-size=4096 %{!shared:%{!static:-dynamic-linker " DYNAMIC_LINKER "}} " \
+    "%{!shared: %{!static: %{rdynamic:-export-dynamic}}}"
 
 #undef TARGET_OS_CPP_BUILTINS
 #define TARGET_OS_CPP_BUILTINS()        \
diff --git a/libgcc/config.host b/libgcc/config.host
index 4d08bed67..77da76555 100644
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -396,8 +396,9 @@ i[34567]86-*-os_2*)
 	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
 	;;
 x86_64-*-os_2*)
-	extra_parts="$extra_parts crti.o crtbegin.o crtend.o crtn.o"
-	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
+	extra_parts="$extra_parts crti.o crtbegin.o crtbeginS.o crtend.o crtendS.o crtn.o"
+	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic t-slibgcc"
+	tm_file="${tm_file} i386/elf-lib.h"
 	;;
 amdgcn*-*-amdhsa)
 	tmake_file="$tmake_file gcn/t-amdgcn"
diff --git a/libitm/configure b/libitm/configure
index fb742d79e..ac8cad442 100644
--- a/libitm/configure
+++ b/libitm/configure
@@ -6735,6 +6735,10 @@ linux* | k*bsd*-gnu | kopensolaris*-gnu)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+os_2*)
+  lt_cv_deplibs_check_method=pass_all
+  ;;
+
 netbsd*)
   if echo __ELF__ | $CC -E - | $GREP __ELF__ > /dev/null; then
     lt_cv_deplibs_check_method='match_pattern /lib[^/]+(\.so\.[0-9]+\.[0-9]+|_pic\.a)$'
@@ -11507,6 +11511,18 @@ solaris*)
   postinstall_cmds='chmod +x $lib'
   ;;
 
+os_2*)
+  version_type=linux
+  need_lib_prefix=no
+  need_version=no
+  library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${release}${shared_ext}$major $libname${shared_ext}'
+  soname_spec='${libname}${release}${shared_ext}$major'
+  shlibpath_var=LD_LIBRARY_PATH
+  shlibpath_overrides_runpath=yes
+  hardcode_into_libs=yes
+  dynamic_linker='os_2 loader.so'
+  ;;
+
 sunos4*)
   version_type=sunos
   library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${shared_ext}$versuffix'
@@ -15154,6 +15170,18 @@ rdos*)
   dynamic_linker=no
   ;;
 
+os_2*)
+  version_type=linux
+  need_lib_prefix=no
+  need_version=no
+  library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${release}${shared_ext}$major $libname${shared_ext}'
+  soname_spec='${libname}${release}${shared_ext}$major'
+  shlibpath_var=LD_LIBRARY_PATH
+  shlibpath_overrides_runpath=yes
+  hardcode_into_libs=yes
+  dynamic_linker='os_2 loader.so'
+  ;;
+
 solaris*)
   version_type=linux
   need_lib_prefix=no
diff --git a/libitm/configure.tgt b/libitm/configure.tgt
index 4c0b60203..81c435588 100644
--- a/libitm/configure.tgt
+++ b/libitm/configure.tgt
@@ -141,7 +141,8 @@ case "${target}" in
   *-*-gnu* | *-*-k*bsd*-gnu \
   | *-*-netbsd* | *-*-freebsd* | *-*-openbsd* \
   | *-*-solaris2* | *-*-sysv4* | *-*-hpux11* \
-  | *-*-darwin* | *-*-aix* | *-*-dragonfly*)
+  | *-*-darwin* | *-*-aix* | *-*-dragonfly* \ 
+  | *-*-os_2*)
 	# POSIX system.  The OS is supported.
 	;;
 
diff --git a/libstdc++-v3/configure b/libstdc++-v3/configure
index 427b4b871..24ed8f0b7 100755
--- a/libstdc++-v3/configure
+++ b/libstdc++-v3/configure
@@ -6626,6 +6626,10 @@ gnu*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
+os_2*)
+  lt_cv_deplibs_check_method=pass_all
+  ;;
+
 haiku*)
   lt_cv_deplibs_check_method=pass_all
   ;;
@@ -11317,6 +11321,18 @@ solaris*)
   postinstall_cmds='chmod +x $lib'
   ;;
 
+os_2*)
+  version_type=linux
+  need_lib_prefix=no
+  need_version=no
+  library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${release}${shared_ext}$major $libname${shared_ext}'
+  soname_spec='${libname}${release}${shared_ext}$major'
+  shlibpath_var=LD_LIBRARY_PATH
+  shlibpath_overrides_runpath=yes
+  hardcode_into_libs=yes
+  dynamic_linker='os_2 loader.so'
+  ;;
+
 sunos4*)
   version_type=sunos
   library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${shared_ext}$versuffix'
@@ -15001,6 +15017,18 @@ solaris*)
   postinstall_cmds='chmod +x $lib'
   ;;
 
+os_2*)
+  version_type=linux
+  need_lib_prefix=no
+  need_version=no
+  library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${release}${shared_ext}$major $libname${shared_ext}'
+  soname_spec='${libname}${release}${shared_ext}$major'
+  shlibpath_var=LD_LIBRARY_PATH
+  shlibpath_overrides_runpath=yes
+  hardcode_into_libs=yes
+  dynamic_linker='os_2 loader.so'
+  ;;
+
 sunos4*)
   version_type=sunos
   library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${shared_ext}$versuffix'
