diff --git a/Makefile b/Makefile
index 4f9a631..d0204f8 100644
--- a/Makefile
+++ b/Makefile
@@ -1,9 +1,9 @@
-CC = gcc
+CC = x86_64-os_2-gcc
 CFLAGS = -O3 -Wall
 
-hw3: mysh.o
-	gcc mysh.o -o mysh
-mysh.o: mysh.c
-	gcc mysh.c -c
+.PHONY: all
+all: mysh
+	$(CC) -o mysh mysh.c $(FLAGS) -T link.ld
+
 clean:
 	rm -rf mysh.o mysh
diff --git a/link.ld b/link.ld
new file mode 100644
index 0000000..a9bfc41
--- /dev/null
+++ b/link.ld
@@ -0,0 +1,45 @@
+OUTPUT_FORMAT("elf64-x86-64")
+OUTPUT_ARCH(i386:x86-64)
+
+/* Sets first thing to be executed to be start */
+ENTRY(_start)
+
+SECTIONS
+{
+    . = 0x400000 + SIZEOF_HEADERS;
+
+    .text :
+    {
+        KEEP(*(SORT_NONE(.init)))
+        *(.text.startup)
+        *(.text .text.*)
+        KEEP(*(SORT_NONE(.fini)))
+    }
+
+    /* Put read only data */
+    .rodata ALIGN(0x1000) :
+    {
+        *(.rodata .rodata.*)
+        *(.eh_frame)
+    }
+
+    /* Put data */
+    .data ALIGN(0x1000) :
+    {
+        *(.data)
+        *(.ctors)
+        *(.dtors)
+    }
+
+    /* Put uninitialized data */
+    .bss ALIGN(0x1000) :
+    {
+        *(COMMON)
+        *(.bss .bss.*)
+    }
+
+    /DISCARD/ : { 
+        *(.comment)
+        *(.debug*)
+    }
+}
\ No newline at end of file
