diff --git a/Makefile b/Makefile
index 13620cc..0a0901f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,7 @@
 all: kilo
 
+CC=x86_64-os_2-gcc
+
 kilo: kilo.c
 	$(CC) -o kilo kilo.c -Wall -W -pedantic -std=c99
 
diff --git a/kilo.c b/kilo.c
index 9490a77..184d45a 100644
--- a/kilo.c
+++ b/kilo.c
@@ -159,7 +159,7 @@ void editorSetStatusMessage(const char *fmt, ...);
  * There is no support to highlight patterns currently. */
 
 /* C / C++ */
-char *C_HL_extensions[] = {".c",".cpp",NULL};
+char *C_HL_extensions[] = {".c",".cpp",".h",NULL};
 char *C_HL_keywords[] = {
         /* A few C / C++ keywords */
         "switch","if","while","for","break","continue","return","else",
@@ -191,6 +191,7 @@ void disableRawMode(int fd) {
     /* Don't even check the return value as it's too late. */
     if (E.rawmode) {
         tcsetattr(fd,TCSAFLUSH,&orig_termios);
+        write(fd, "\033[1;1H\033[3J", 10);
         E.rawmode = 0;
     }
 }
